File: app/src/main/java/com/example/numad25sp_wenyupan_java/CalculatorActivity.java
package com.example.numad25sp_wenyupan_java;

import android.os.Bundle;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import androidx.appcompat.app.AppCompatActivity;

public class CalculatorActivity extends AppCompatActivity {

    private TextView calcDisplay;
    private StringBuilder currentInput = new StringBuilder();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_calculator);

        calcDisplay = findViewById(R.id.textView);

        setButtonClickListeners();
    }

    private void setButtonClickListeners() {
        int[] buttonIds = {
                R.id.btn0, R.id.btn1, R.id.btn2, R.id.btn3, R.id.btn4,
                R.id.btn5, R.id.btn6, R.id.btn7, R.id.btn8, R.id.btn9,
                R.id.btnPlus, R.id.buttonMinus
        };

        View.OnClickListener numberClickListener = new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Button clickedButton = (Button) v;
                currentInput.append(clickedButton.getText().toString());
                calcDisplay.setText(currentInput.toString()); // Update display
            }
        };

        for (int id : buttonIds) {
            findViewById(id).setOnClickListener(numberClickListener);
        }

        findViewById(R.id.btnDivide).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (currentInput.length() > 0) {
                    String result = evaluateMathExpression(currentInput.toString());
                    calcDisplay.setText(result);
                    currentInput.setLength(0);
                    if (!result.equals("Error")) {
                        currentInput.append(result);
                    }
                }
            }
        });

        findViewById(R.id.btnDelete).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (currentInput.length() > 0) {
                    currentInput.deleteCharAt(currentInput.length() - 1);
                    calcDisplay.setText(currentInput.length() > 0 ? currentInput.toString() : "CALC");
                }
            }
        });
    }

    private String evaluateMathExpression(String expression) {
        try {
            return String.valueOf(evaluate(expression));
        } catch (Exception e) {
            return "Error";
        }
    }

    private int evaluate(String expression) {
        int result = 0;
        int currentNumber = 0;
        int sign = 1; // 1 for '+', -1 for '-'

        for (int i = 0; i < expression.length(); i++) {
            char c = expression.charAt(i);

            if (Character.isDigit(c)) {
                currentNumber = currentNumber * 10 + (c - '0');
            } else if (c == '+' || c == '-') {
                result += sign * currentNumber;
                currentNumber = 0;
                sign = (c == '+') ? 1 : -1;
            }
        }
        result += sign * currentNumber;

        return result;
    }
}

File: app/src/main/java/com/example/numad25sp_wenyupan/CalculatorActivity.kt
package com.example.numad25sp_wenyupan

import android.os.Bundle
import android.widget.Button
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import net.objecthunter.exp4j.ExpressionBuilder

class CalculatorActivity : AppCompatActivity() {

    private lateinit var display: TextView
    private var currentExpression: String = ""

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_calculator) // Link to your XML file

        display = findViewById(R.id.displayText)

        val buttons = listOf(
            R.id.button0, R.id.button1, R.id.button2, R.id.button3, R.id.button4,
            R.id.button5, R.id.button6, R.id.button7, R.id.button8, R.id.button9,
            R.id.buttonPlus, R.id.buttonMinus, R.id.buttonMultiply, R.id.buttonEqual
        )

        for (id in buttons) {
            findViewById<Button>(id).setOnClickListener { buttonClicked(it as Button) }
        }
    }

    private fun buttonClicked(button: Button) {
        val value = button.text.toString()
        if (value == "=") {
            try {
                currentExpression = evaluateExpression(currentExpression).toString()
            } catch (e: Exception) {
                currentExpression = "Error"
            }
        } else {
            currentExpression += value
        }
        display.text = currentExpression
    }

    private fun evaluateExpression(expression: String): Double {
        return ExpressionBuilder(expression).build().evaluate()
    }
}
